File: asset-web-backend/app/routers/court.py
"""
法院卷宗模块 - API 路由
"""
from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import FileResponse, StreamingResponse
from typing import Optional, List
import os
import mimetypes

from app.models.court import (
    CaseSearchRequest, CaseSearchResponse,
    DossierListResponse, CaseDetail,
    DOSSIER_CATEGORY_MAP
)
from app.services.court_service import court_service

router = APIRouter(prefix="/api/court", tags=["法院卷宗"])


@router.post("/cases/search", response_model=CaseSearchResponse)
async def search_cases(request: CaseSearchRequest):
    """
    检索案件列表
    
    支持按案号、案由、案件类型、审判阶段、承办人、日期范围筛选
    """
    cases, total = await court_service.search_cases(
        ah=request.ah,
        ay_ms=request.ay_ms,
        ajlx_mc=request.ajlx_mc,
        trial_stage=request.trial_stage,
        cbr_mc=request.cbr_mc,
        region=request.region,
        dossier_categories=request.dossier_categories,
        start_date=request.start_date,
        end_date=request.end_date,
        keyword=request.keyword,
        page=request.page,
        page_size=request.page_size
    )
    
    return CaseSearchResponse(
        total=total,
        page=request.page,
        page_size=request.page_size,
        items=cases
    )


@router.get("/regions")
async def get_regions():
    """获取所有区域列表（从数据库动态获取）"""
    regions = await court_service.get_regions()
    return {"regions": regions}


@router.get("/cases/{case_code}/download-zip")
async def download_case_dossiers_zip(case_code: str):
    """打包下载案件所有卷宗（ZIP格式）"""
    import zipfile
    from io import BytesIO
    
    # 获取案件所有卷宗文件路径
    dossiers = await court_service.get_case_dossier_files(case_code)
    if not dossiers:
        raise HTTPException(status_code=404, detail="未找到该案件的卷宗文件")
    
    # 创建内存中的ZIP文件
    zip_buffer = BytesIO()
    with zipfile.ZipFile(zip_buffer, 'w', zipfile.ZIP_DEFLATED) as zip_file:
        for dossier in dossiers:
            filepath = dossier.get('filepath')
            if filepath and os.path.exists(filepath):
                # 使用原始文件名作为ZIP内的文件名
                arcname = dossier.get('original_name', os.path.basename(filepath))
                zip_file.write(filepath, arcname)
    
    zip_buffer.seek(0)
    
    # 返回ZIP文件
    return StreamingResponse(
        zip_buffer,
        media_type="application/zip",
        headers={
            "Content-Disposition": f'attachment; filename="{case_code}.zip"'
        }
    )


@router.get("/cases/{case_code}", response_model=CaseDetail)
async def get_case_detail(case_code: str):
    """获取案件详情"""
    case = await court_service.get_case_detail(case_code)
    if not case:
        raise HTTPException(status_code=404, detail="案件不存在")
    return case


@router.get("/cases/{case_code}/dossiers", response_model=DossierListResponse)
async def get_case_dossiers(case_code: str):
    """
    获取案件的卷宗列表
    
    返回平铺列表和树形结构两种格式
    """
    # 先验证案件存在
    case = await court_service.get_case_detail(case_code)
    if not case:
        raise HTTPException(status_code=404, detail="案件不存在")
    
    dossiers = await court_service.get_dossiers(case_code)
    tree = court_service.build_dossier_tree(dossiers)
    
    return DossierListResponse(
        case_code=case_code,
        total=len(dossiers),
        items=dossiers,
        tree=tree
    )


@router.get("/dossiers/{dossier_code}/download")
async def download_dossier(dossier_code: str):
    """下载卷宗文件"""
    filepath = await court_service.get_dossier_filepath(dossier_code)
    if not filepath:
        raise HTTPException(status_code=404, detail="卷宗文件不存在")
    
    if not os.path.exists(filepath):
        raise HTTPException(status_code=404, detail="文件不存在于存储中")
    
    filename = os.path.basename(filepath)
    
    return FileResponse(
        path=filepath,
        filename=filename,
        media_type="application/octet-stream"
    )


@router.get("/case-types")
async def get_case_types():
    """获取案件类型列表（用于下拉筛选）"""
    types = await court_service.get_case_types()
    return {"items": types}


@router.get("/case-reasons")
async def get_case_reasons():
    """获取案由列表（用于下拉筛选）"""
    reasons = await court_service.get_case_reasons()
    return {"items": reasons}


@router.get("/dossier-categories")
async def get_dossier_categories():
    """获取卷宗分类列表"""
    return {"items": [
        {"id": k, "name": v} for k, v in DOSSIER_CATEGORY_MAP.items()
    ]}


@router.get("/dossiers/search")
async def search_dossiers(
    category_ids: Optional[str] = Query(None, description="分类ID，多个用逗号分隔"),
    case_code: Optional[str] = Query(None, description="案件编号"),
    keyword: Optional[str] = Query(None, description="关键词"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量")
):
    """
    按分类搜索卷宗
    
    - category_ids: 分类ID，多个用逗号分隔，如 "1,3,6"
    - case_code: 案件编号（模糊搜索）
    - keyword: 文件名关键词
    """
    # 解析分类ID列表
    category_list = None
    if category_ids:
        try:
            category_list = [int(x.strip()) for x in category_ids.split(',')]
        except ValueError:
            raise HTTPException(status_code=400, detail="category_ids 格式错误")
    
    dossiers, total = await court_service.search_dossiers_by_category(
        category_ids=category_list,
        case_code=case_code,
        keyword=keyword,
        page=page,
        page_size=page_size
    )
    
    return {
        "total": total,
        "page": page,
        "page_size": page_size,
        "items": dossiers
    }


@router.get("/dossiers/category-stats")
async def get_category_stats():
    """获取各分类的卷宗统计"""
    stats = await court_service.get_category_stats()
    return {"items": stats}

@router.get("/dossiers/{dossier_code}/preview")
async def preview_dossier(dossier_code: str):
    """在线预览卷宗文件（支持PDF、图片等）"""
    filepath = await court_service.get_dossier_filepath(dossier_code)
    if not filepath:
        raise HTTPException(status_code=404, detail="卷宗文件不存在")
    
    if not os.path.exists(filepath):
        raise HTTPException(status_code=404, detail="文件不存在于存储中")
    
    # 根据文件扩展名确定 MIME 类型
    ext = os.path.splitext(filepath)[1].lower()
    
    # 可以在浏览器内显示的类型
    previewable_types = {
        '.pdf': 'application/pdf',
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
        '.png': 'image/png',
        '.gif': 'image/gif',
        '.bmp': 'image/bmp',
        '.txt': 'text/plain; charset=utf-8',
    }
    
    mime_type = previewable_types.get(ext)
    
    if mime_type:
        # 可预览类型：使用 Response 类手动返回，避免 FileResponse 自动设置 attachment
        from urllib.parse import quote
        from starlette.responses import Response
        
        filename = os.path.basename(filepath)
        encoded_filename = quote(filename, safe='')
        
        # 读取文件内容
        with open(filepath, 'rb') as f:
            content = f.read()
        
        return Response(
            content=content,
            media_type=mime_type,
            headers={
                "Content-Disposition": f"inline; filename*=UTF-8''{encoded_filename}"
            }
        )
    else:
        # 不可预览类型：触发下载
        from urllib.parse import quote
        filename = os.path.basename(filepath)
        encoded_filename = quote(filename, safe='')
        return FileResponse(
            path=filepath,
            media_type='application/octet-stream',
            headers={
                "Content-Disposition": f"attachment; filename*=UTF-8''{encoded_filename}"
            }
        )

File: asset-web-frontend/src/api/court.js
/**
 * 法院卷宗模块 API
 */
import axios from 'axios'

const API_BASE = '/api/court'

/**
 * 检索案件列表
 * @param {Object} params - 搜索参数
 */
export async function searchCases(params) {
    const response = await axios.post(`${API_BASE}/cases/search`, {
        ah: params.ah || null,
        ay_ms: params.ay_ms || null,
        ajlx_mc: params.ajlx_mc || null,
        trial_stage: params.trial_stage || null,
        cbr_mc: params.cbr_mc || null,
        region: params.region || null,
        dossier_categories: params.dossier_categories && params.dossier_categories.length > 0
            ? params.dossier_categories : null,
        start_date: params.start_date || null,
        end_date: params.end_date || null,
        keyword: params.keyword || null,
        page: params.page || 1,
        page_size: params.page_size || 20
    })
    return response.data
}

/**
 * 获取所有区域列表（动态从数据库获取）
 */
export async function getRegions() {
    const response = await axios.get(`${API_BASE}/regions`)
    return response.data.regions
}

/**
 * 获取案件详情
 * @param {string} caseCode - 案件编码
 */
export async function getCaseDetail(caseCode) {
    const response = await axios.get(`${API_BASE}/cases/${caseCode}`)
    return response.data
}

/**
 * 获取案件卷宗列表
 * @param {string} caseCode - 案件编码
 */
export async function getCaseDossiers(caseCode) {
    const response = await axios.get(`${API_BASE}/cases/${caseCode}/dossiers`)
    return response.data
}

/**
 * 获取卷宗预览URL（浏览器内显示）
 * @param {string} dossierCode - 卷宗编码
 */
export function getDossierPreviewUrl(dossierCode) {
    return `${API_BASE}/dossiers/${encodeURIComponent(dossierCode)}/preview`
}

/**
 * 获取卷宗下载URL
 * @param {string} dossierCode - 卷宗编码
 */
export function getDossierDownloadUrl(dossierCode) {
    return `${API_BASE}/dossiers/${dossierCode}/download`
}

/**
 * 下载卷宗文件
 * @param {string} dossierCode - 卷宗编码
 * @param {string} filename - 文件名
 */
export async function downloadDossier(dossierCode, filename) {
    const url = getDossierDownloadUrl(dossierCode)
    const link = document.createElement('a')
    link.href = url
    link.download = filename
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
}

/**
 * 获取案件类型列表
 */
export async function getCaseTypes() {
    const response = await axios.get(`${API_BASE}/case-types`)
    return response.data.items
}

/**
 * 获取案由列表
 */
export async function getCaseReasons() {
    const response = await axios.get(`${API_BASE}/case-reasons`)
    return response.data.items
}

/**
 * 获取卷宗分类列表
 */
export async function getDossierCategories() {
    const response = await axios.get(`${API_BASE}/dossier-categories`)
    return response.data.items
}

/**
 * 按分类搜索卷宗
 * @param {Object} params - 搜索参数
 * @param {Array} params.category_ids - 分类ID列表
 * @param {string} params.case_code - 案件编号
 * @param {string} params.keyword - 关键词
 * @param {number} params.page - 页码
 * @param {number} params.page_size - 每页数量
 */
export async function searchDossiersByCategory(params) {
    const queryParams = new URLSearchParams()

    if (params.category_ids && params.category_ids.length > 0) {
        queryParams.append('category_ids', params.category_ids.join(','))
    }
    if (params.case_code) {
        queryParams.append('case_code', params.case_code)
    }
    if (params.keyword) {
        queryParams.append('keyword', params.keyword)
    }
    queryParams.append('page', params.page || 1)
    queryParams.append('page_size', params.page_size || 20)

    const response = await axios.get(`${API_BASE}/dossiers/search?${queryParams.toString()}`)
    return response.data
}

/**
 * 获取分类统计
 */
export async function getCategoryStats() {
    const response = await axios.get(`${API_BASE}/dossiers/category-stats`)
    return response.data.items
}

/**
 * 获取案件卷宗 ZIP 打包下载 URL
 * @param {string} caseCode - 案件编号
 */
export function getCaseZipDownloadUrl(caseCode) {
    return `${API_BASE}/cases/${encodeURIComponent(caseCode)}/download-zip`
}
